File: solutions/0004/0004.c
#include <stdio.h>

static int IsPalindrome( int Num )
{
  int Orig = Num;
  int Reverse = Num%10;
  Num /= 10;

  while ( Num )
  {
    Reverse = Reverse*10 + Num%10;
    Num /= 10;
  }

  return Orig == Reverse;
}

int main()
{
  int LargestPalindrome = 0;
  int Product;
  
  for ( int a = 999; a >= 100; a-- )
  {
    for ( int b = 999; b >= a; b-- )
    {
      Product = a*b;

      if ( Product <= LargestPalindrome )
        break;

      if ( IsPalindrome( Product ) )
        LargestPalindrome = Product;
    }
  }
   
  printf( "%d\n", LargestPalindrome );
}

File: solutions/0004/0004.py
def is_palindrome(num):
    orig = num
    rev = num%10
    num //= 10

    while num:
        rev = rev*10 + num%10
        num //= 10

    return orig == rev

largest = 0

for a in range(999, 99, -1):
    for b in range (999, a, -1):
        prod = a*b
        
        if prod < largest:
            break;

        if is_palindrome(prod):
            largest = prod

print(largest)
